File: Planarian/Planarian/Modules/TripObjectives/Services/TripObjectiveService.cs
using Planarian.Library.Helpers;
using Planarian.Model.Database.Entities;
using Planarian.Model.Database.Entities.TripObjectives;
using Planarian.Model.Shared;
using Planarian.Modules.Leads.Models;
using Planarian.Modules.Project.Controllers;
using Planarian.Modules.TripObjectives.Controllers;
using Planarian.Modules.TripObjectives.Repositories;
using Planarian.Modules.Users.Repositories;
using Planarian.Shared.Base;
using Planarian.Shared.Services;

namespace Planarian.Modules.TripObjectives.Services;

public class TripObjectiveService : ServiceBase<TripObjectiveRepository>
{
    private readonly BlobService _blobService;
    private readonly TagRepository _tagRepository;
    private readonly UserRepository _userRepository;

    public TripObjectiveService(TripObjectiveRepository repository, RequestUser requestUser, BlobService blobService,
        TagRepository tagRepository, UserRepository userRepository) :
        base(repository, requestUser)
    {
        _blobService = blobService;
        _tagRepository = tagRepository;
        _userRepository = userRepository;
    }

    #region Photos

    public async Task UploadPhotos(IEnumerable<TripPhotoUpload> photos, string tripObjectiveId)
    {
        var ids = await Repository.GetIds(tripObjectiveId);
        if (ids == null) throw new ArgumentNullException(nameof(tripObjectiveId));

        foreach (var photo in photos)
        {
            var title = !string.IsNullOrWhiteSpace(photo.Title) ? photo.Title : photo.File.FileName;

            var fileType = Path.GetExtension(photo.File.FileName);

            if (!FileValidation.IsValidPhotoFileType(fileType)) continue;
            // TODO alert user
            var entity = new TripPhoto(RequestUser.Id, tripObjectiveId, title, photo.Description, fileType);
            Repository.Add(entity);
            await Repository.SaveChangesAsync();
            var blobKey = await _blobService.AddTripPhoto(ids.ProjectId, ids.TripId, tripObjectiveId, entity.Id,
                photo.File.OpenReadStream(), fileType);
            entity.BlobKey = blobKey;
            await Repository.SaveChangesAsync();
        }
    }

    #endregion

    public async Task<IEnumerable<SelectListItem<string>>> GetTripObjectiveMembers(string tripObjectiveId)
    {
        var tripObjectiveMembers = await Repository.GetTripObjectiveMembers(tripObjectiveId);

        return tripObjectiveMembers;
    }

    public async Task AddOrUpdateTripReport(string tripObjectiveId, string tripReport)
    {
        var tripObjective = await Repository.GetTripObjective(tripObjectiveId);

        if (tripObjective == null) throw new NullReferenceException("Trip objective does not exist");

        tripObjective.TripReport = tripReport;
        await Repository.SaveChangesAsync();
    }

    public async Task<IEnumerable<TripPhotoVm>> GetPhotos(string tripObjectiveId)
    {
        var photos = (await Repository.GetTripObjectivePhotos(tripObjectiveId)).ToList();

        foreach (var photo in photos)
        {
            var uri = _blobService.GetSasUrl(photo.Url, 24);
            if (uri?.AbsolutePath != null) photo.Url = uri.AbsoluteUri;
        }

        return photos;
    }

    public async Task UpdateObjectiveName(string tripObjectiveId, string name)
    {
        var entity = await Repository.GetTripObjective(tripObjectiveId);

        if (entity == null) throw new NullReferenceException("Trip objective does not exist");

        entity.Name = name;
        await Repository.SaveChangesAsync();
    }

    public async Task UpdateObjectiveDescription(string tripObjectiveId, string description)
    {
        var entity = await Repository.GetTripObjective(tripObjectiveId);

        if (entity == null) throw new NullReferenceException("Trip objective does not exist");

        entity.Description = description;
        await Repository.SaveChangesAsync();
    }


    public async Task InviteTripObjectiveMember(string tripObjectiveId, InviteMember invitation)
    {
        var tripObjective = await Repository.GetTripObjective(tripObjectiveId);

        if (tripObjective == null) throw new NullReferenceException("Trip objective not found");

        var user = await _userRepository.GetUserByEmail(invitation.Email);
        if (user != null)
        {
            await AddTripObjectiveMember(tripObjectiveId, user.Id);
        }
        else
        {
            var entity = new User(invitation.FirstName, invitation.LastName, invitation.Email);
            _userRepository.Add(entity);
            await Repository.SaveChangesAsync();
            await AddTripObjectiveMember(tripObjectiveId, entity.Id);
        }
    }

    #region Project

    public async Task<TripObjectiveVm> CreateOrUpdateTripObjective(CreateOrEditTripObjectiveVm values)
    {
        var isNew = string.IsNullOrWhiteSpace(values.Id);
        var tripObjective = values.Id != null
            ? await Repository.GetTripObjective(values.Id) ?? new TripObjective()
            : new TripObjective();

        tripObjective.TripId = values.TripId;

        foreach (var tripObjectiveTypeId in values.TripObjectiveTypeIds)
        {
            var tripObjectiveType = Repository.GetTripObjectiveType(tripObjectiveTypeId);
            if (tripObjectiveType == null) throw new NullReferenceException("Tag not found");

            var tripObjectiveTag = new TripObjectiveTag
            {
                Tag = tripObjectiveType
            };
            tripObjective.TripObjectiveTags.Add(tripObjectiveTag);
        }

        tripObjective.Name = values.Name;
        tripObjective.Description = values.Description;
        tripObjective.TripReport = values.TripReport;

        if (isNew)
        {
            Repository.Add(tripObjective);
            await Repository.SaveChangesAsync();
        }

        foreach (var tripObjectiveMemberId in values.TripObjectiveMemberIds)
            await AddTripObjectiveMember(tripObjective.Id, tripObjectiveMemberId, false);

        await Repository.SaveChangesAsync();
        return new TripObjectiveVm(tripObjective, values.TripObjectiveTypeIds, values.TripObjectiveMemberIds);
    }

    public async Task<TripObjectiveVm?> GetTripObjective(string projectId)
    {
        return await Repository.GetTripObjectiveVm(projectId);
    }

    public async Task DeleteTripObjective(string projectId)
    {
        var tripObjective = await Repository.GetTripObjective(projectId);
        if (tripObjective != null)
        {
            Repository.Delete(tripObjective);
            await Repository.SaveChangesAsync();
        }
    }

    #endregion

    #region Trip Objective Member

    public async Task AddTripObjectiveMember(string tripObjectiveId, string userId, bool saveChanges = true)
    {
        await AddTripObjectiveMember(tripObjectiveId, new List<string> { userId }, saveChanges);
    }

    public async Task AddTripObjectiveMember(string tripObjectiveId, IEnumerable<string> userIds,
        bool saveChanges = true)
    {
        var tripObjective = await Repository.GetTripObjective(tripObjectiveId);
        if (tripObjective == null) throw new NullReferenceException("Trip Objective not found");

        foreach (var userId in userIds)
        {
            var tripObjectiveMember = new TripObjectiveMember
            {
                TripObjective = tripObjective,
                UserId = userId
            };
            tripObjective.TripObjectiveMembers.Add(tripObjectiveMember);
        }

        if (saveChanges) await Repository.SaveChangesAsync();
    }

    public async Task DeleteTripObjectiveMember(string tripObjectiveId, string userId)
    {
        var projectMember = await Repository.GetTripObjectiveMember(tripObjectiveId, userId);
        if (projectMember != null)
        {
            Repository.Delete(projectMember);
            await Repository.SaveChangesAsync();
        }
    }

    #endregion

    #region Leads

    public async Task<IEnumerable<LeadVm>> GetLeads(string tripObjectiveId)
    {
        return await Repository.GetLeads(tripObjectiveId);
    }

    public async Task AddLeads(IEnumerable<CreateLeadVm> leads, string tripObjectiveId)
    {
        foreach (var lead in leads)
        {
            var entity = new Lead(lead, tripObjectiveId, RequestUser.Id);
            Repository.Add(entity);
        }

        await Repository.SaveChangesAsync();
    }

    #endregion

    #region Tags

    public async Task<IEnumerable<SelectListItem<string>>> GetTripObjectiveTags(string tripObjectiveId)
    {
        return await Repository.GetTripObjectiveTags(tripObjectiveId);
    }

    public async Task AddTagType(string tagId, string tripObjectiveId)
    {
        var tagType = await _tagRepository.GetTag(tagId);

        if (tagType == null) throw new NullReferenceException("Tag type does not exist");

        var tripObjective = await Repository.GetTripObjective(tripObjectiveId);

        if (tripObjective == null) throw new NullReferenceException("Trip objective does not exist");

        var tripObjectiveTag = new TripObjectiveTag
        {
            TripObjectiveId = tripObjectiveId,
            Tag = tagType
        };
        tripObjective.TripObjectiveTags.Add(tripObjectiveTag);

        await Repository.SaveChangesAsync();
    }

    public async Task DeleteTripObjectiveTag(string tagId, string tripObjectiveId)
    {
        var tag = await Repository.GetTripObjectiveTag(tagId, tripObjectiveId);

        if (tag == null) throw new NullReferenceException("Tag does not exist");

        Repository.Delete(tag);

        await Repository.SaveChangesAsync();
    }

    #endregion
}

File: Planarian.Web/src/Modules/Objective/Services/trip.objective.service.ts
import { RcFile } from "antd/lib/upload";
import { HttpClient } from "../../..";
import { InviteMember } from "../../../Shared/Components/InviteMember";
import { HttpHelpers } from "../../../Shared/Helpers/HttpHelpers";
import { isNullOrWhiteSpace } from "../../../Shared/Helpers/StringHelpers";
import { SelectListItem } from "../../../Shared/Models/SelectListItem";
import { CreateLeadVm } from "../../Leads/Models/CreateLeadVm";
import { LeadVm } from "../../Leads/Models/Lead";
import { CreateOrEditTripObjectiveVm } from "../Models/CreateOrEditTripObjectiveVm";
import { TripObjectiveVm } from "../Models/TripObjectiveVm";
import { TripPhotosUpload, TripPhotoUpload } from "../Models/TripPhotoUpload";
import { TripPhotoVm } from "../Models/TripPhotoVm";

const baseUrl = "api/tripObjectives";
const TripObjectiveService = {
  //#region Objectives
  async GetObjective(id: string): Promise<TripObjectiveVm> {
    const response = await HttpClient.get<TripObjectiveVm>(`${baseUrl}/${id}`);
    return response.data;
  },
  async GetTags(id: string): Promise<SelectListItem<string>[]> {
    const response = await HttpClient.get<SelectListItem<string>[]>(
      `${baseUrl}/${id}/tags`
    );
    return response.data;
  },
  async AddTag(
    tagId: string,
    tripObjectiveId: string
  ): Promise<SelectListItem<string>[]> {
    const response = await HttpClient.post<SelectListItem<string>[]>(
      `${baseUrl}/${tripObjectiveId}/tags/${tagId}`,
      {}
    );
    return response.data;
  },
  async DeleteTag(
    tagId: string,
    tripObjectiveId: string
  ): Promise<SelectListItem<string>[]> {
    const response = await HttpClient.delete<SelectListItem<string>[]>(
      `${baseUrl}/${tripObjectiveId}/tags/${tagId}`
    );
    return response.data;
  },
  async AddTripObjective(
    values: CreateOrEditTripObjectiveVm
  ): Promise<TripObjectiveVm> {
    const response = await HttpClient.post<TripObjectiveVm>(
      `${baseUrl}`,
      values
    );
    return response.data;
  },

  async AddOrUpdateTripReport(
    tripObjectiveId: string,
    tripReport: string
  ): Promise<void> {
    const response = await HttpClient.post(
      `${baseUrl}/${tripObjectiveId}/tripReport`,
      tripReport,
      { headers: { "Content-Type": "application/json" } }
    );
  },

  async UpdateObjectiveName(
    name: string,
    tripObjectiveId: string
  ): Promise<void> {
    const response = await HttpClient.post(
      `${baseUrl}/${tripObjectiveId}/name`,
      name,
      { headers: { "Content-Type": "application/json" } }
    );
  },
  async UpdateObjectiveDescription(
    description: string,
    tripObjectiveId: string
  ): Promise<void> {
    const response = await HttpClient.post(
      `${baseUrl}/${tripObjectiveId}/description`,
      description,
      { headers: { "Content-Type": "application/json" } }
    );
  },
  //#endregion

  //#region Photos
  async UploadPhotos(
    values: TripPhotoUpload[],
    tripObjectiveId: string
  ): Promise<void> {
    const formData = new FormData();

    values.forEach((e, i) => {
      const title = values[i].title;
      const description = values[i].description;
      formData.append(
        `formData[${i}].title`,
        isNullOrWhiteSpace(title) ? "" : (title as string)
      );
      formData.append(
        `formData[${i}].description`,
        isNullOrWhiteSpace(description) ? "" : (description as string)
      );
      formData.append(`formData[${i}].file`, values[i].file as RcFile);
      formData.append(`formData[${i}].uid`, values[i].uid as string);
    });

    const response = await HttpClient.post<void>(
      `${baseUrl}/${tripObjectiveId}/photos`,
      formData,
      { headers: { "Content-Type": "multipart/form-data" } }
    );
    return response.data;
  },

  async GetTripObjectivePhotos(objectiveId: string): Promise<TripPhotoVm[]> {
    const response = await HttpClient.get<TripPhotoVm[]>(
      `${baseUrl}/${objectiveId}/photos`
    );
    return response.data;
  },

  //#endregion

  //#region Trip Objective Members
  async GetTripObjectiveMembers(
    objectiveId: string
  ): Promise<SelectListItem<string>[]> {
    const response = await HttpClient.get<SelectListItem<string>[]>(
      `${baseUrl}/${objectiveId}/members`
    );
    return response.data;
  },
  async AddTripObjectiveMember(
    userId: string,
    objectiveId: string
  ): Promise<SelectListItem<string>[]> {
    const response = await HttpClient.post<SelectListItem<string>[]>(
      `${baseUrl}/${objectiveId}/members/${userId}`,
      {}
    );
    return response.data;
  },
  async AddTripObjectiveMembers(
    userIds: string[],
    objectiveId: string
  ): Promise<SelectListItem<string>[]> {
    const response = await HttpClient.post<SelectListItem<string>[]>(
      `${baseUrl}/${objectiveId}/members`,
      userIds
    );
    return response.data;
  },

  async DeleteTripObjectiveMember(
    userId: string,
    objectiveId: string
  ): Promise<SelectListItem<string>[]> {
    const response = await HttpClient.delete<SelectListItem<string>[]>(
      `${baseUrl}/${objectiveId}/members/${userId}`,
      {}
    );
    return response.data;
  },
  //#endregion

  //#region Leads
  async GetLeads(objectiveId: string): Promise<LeadVm[]> {
    const response = await HttpClient.get<LeadVm[]>(
      `${baseUrl}/${objectiveId}/leads`
    );
    return response.data;
  },

  async AddLeads(leads: CreateLeadVm[], objectiveId: string): Promise<void> {
    const response = await HttpClient.post<void>(
      `${baseUrl}/${objectiveId}/leads`,
      leads
    );
  },

  //#endregion

  //#region

  async InviteTripObjectiveMembers(
    values: InviteMember,
    objectiveId: string
  ): Promise<void> {
    const response = await HttpClient.post<void>(
      `${baseUrl}/${objectiveId}/members/invite`,
      values
    );
    return response.data;
  },
  //#endregion
};

export { TripObjectiveService };
